### PR TITLE
Add AppDevAnalytics

### PR DESCRIPTION
**Pull request by camruss**
## Overview

Adds AppDevAnalytics :)

## Changes Made

Logs events made by user as detailed in `Volume/Analytics/Events.md`

## Test Coverage

Manual testing

## Next Steps

Track analytics once in production

### CODE REVIEW

File: Volume/Analytics/Analytics.swift [hunk 1]
```diff
@@ -9,11 +9,82 @@
...
     let name = "start_onboarding"
 }
 
 struct CompleteOnboarding: Event {
     let name = "complete_onboarding"
 }
+
+struct FollowPublication: Event {
+    let name = "follow_publication"
+    var parameters: [String : Any]?
```
Comment (danielVebman) [inline Volume/Analytics/Analytics.swift:22]: You define this exact field 8 different times. To reduce repetition, I would create a base protocol with this value. For example (with a better name)

```
protocol BaseEvent: Event {
    var parameters: [String: Any]?
}
```

Also, do we do `[String : Any]` or `[String: Any]` throughout the codebase? It should be consistent. If I recall correctly, Google's style guide has the latter, but you should confirm if you're unsure.

File: Volume/View Models/ArticleRow.swift [hunk 2]
```diff
@@ -11,15 +11,17 @@ import SwiftUI
 
 struct ArticleRow: View {
     let article: Article
+    let entryPoint: EntryPoint
```
Comment (danielVebman) [inline Volume/View Models/ArticleRow.swift:14]: I like `NavigationSource` just a little more, but that's totally subjective and I like this name too.

File: Volume/Views/MainView/PublicationList.swift [hunk 3]
```diff
@@ -79,7 +79,7 @@ struct PublicationList: View {
                     case .results(let results):
                         LazyHStack(spacing: 12) {
                             ForEach(results.followedPublications) { publication in
-                                NavigationLink(destination: PublicationDetail(publication: publication)) {
+                                NavigationLink(destination: PublicationDetail(entryPoint: .followingPublications, publication: publication)) {
```
Comment (danielVebman) [inline Volume/Views/MainView/PublicationList.swift:82]: You put entry point after the article/publication in other places. Keep it consistent. 

For example, see BookmarksList line 69. I like it with entry point after the model (with the idea of putting the most important parameter first). That consistency outweighs alphabetizing or adding an explicit initializer in my opinion.

File: Volume/Analytics/Analytics.swift [hunk 4]
```diff
@@ -9,11 +9,82 @@
...
+}
+
+struct UnbookmarkArticle: Event {
+    let name = "unbookmark_article"
+    var parameters: [String : Any]?
+}
+
+// Helper struct and enums to disambiguate parameters for different event types
+struct Parameters {
+    static func params(for event: EventLevel, id: String, at entry: EntryPoint) -> [String: Any] {
```
Comment (danielVebman) [inline Volume/Analytics/Analytics.swift:62]: Could be more concise/idiomatic as 

```suggestion
    static func for(_ event: EventLevel, id: String, at entry: EntryPoint) -> [String: Any] {
```

And then call with

```
Parameters.for(.article, id: id, entry: .articleDetail)
```

File: Volume/Analytics/Analytics.swift [hunk 5]
```diff
@@ -9,11 +9,82 @@
...
+    case trendingArticles = "trending_articles"
+    
+    // Publication Entry Points
+    case articleDetail = "article_detail"
+    case followingPublications = "following_publications"
+    case morePublications = "more_publications"
+    case onboarding = "onboarding"
+}
+
+enum EventLevel {
```
Comment (danielVebman) [inline Volume/Analytics/Analytics.swift:87]: This feels a little like a misnomer. Are articles or publications a "level?" Levels imply a hierarchy, which I don't think is present here. They are more like the type, subject, object, etc. of the event. I would play around with this.

**Review by danielVebman — APPROVED** [review]
This is great 🤯!! Just a couple comments. Conner's going to love all the data!

File: Volume/Views/PublicationDetail/PublicationDetailHeader.swift [hunk 6]
```diff
@@ -77,14 +79,18 @@ struct PublicationDetailHeader: View {
                 .padding([.bottom, .top], 8)
             Text(publication.bio)
                 .font(.helveticaRegular(size: 14))
-                .fixedSize(horizontal: false, vertical: true)
-
```
Comment (spd75) [inline Volume/Views/PublicationDetail/PublicationDetailHeader.swift:81]: Why did you get rid of this?

**Review by spd75 — APPROVED** [review]
Looks good!  I agree a lot with Daniel about making a base event.

**Review by cswenberg — APPROVED** [review]
yay this looks great!! super hyped :)

File: Volume/Views/BrowserView.swift [hunk 7]
```diff
@@ -14,11 +15,21 @@ struct BrowserView: View {
...
     var isShoutoutsButtonEnabled: Bool {
         userData.canIncrementShoutouts(article)
     }
 
+    private var params: [String: Any] {
+        Parameters.params(for: .article, id: article.id, at: entryPoint)
+    }
+
     private func incrementShoutouts() {
+        if isShoutoutsButtonEnabled {
```
Comment (danielVebman) [inline Volume/Views/BrowserView.swift:29]: You don't need this check. The function can only be called if the button is enabled. Also, it doesn't make sense to only put the event in here but not the actual incrementing network request.

File: Volume/Analytics/Analytics.swift (hunk 1 above)
Comment (danielVebman) [reply Volume/Analytics/Analytics.swift:22]: Actually I change my mind. I don't think it's possible to omit the field name here unless you add extra code, which defeats the purpose. I would take a different approach:

```
struct AnyEvent: Event {
    let name: String
    let parameters: [String: Any]? 
    // make these vars if necessary 
}

enum VolumeEvent { // or whatever
    case startOnboarding = "start_onboarding"
    // ...

    func toEvent(parameters: [...]) {
        AnyEvent(name: rawValue, parameters: parameters)
    }
}
```

File: Volume/Analytics/Analytics.swift (hunk 4 above)
Comment (camruss) [reply Volume/Analytics/Analytics.swift:62]: for is reserved

**Pull request merged**